File: 030_Substring_with_Concatenation_of_All_Words/030_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
#include <sstream>
#include <set>
using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if (s.empty() || words.empty()) return {};
        vector<int> res;
        size_t one_word = words[0].size();
        size_t word_num = words.size();
        size_t total_len = one_word * word_num;
        unordered_map<string, int> m1;
        for (const auto &w : words) m1[w]++;

        for (int i = 0; i < one_word; i++) {
            unordered_map<string, int> m2;
            int count = 0, left = i, right = i;
            while (right + one_word <= s.size()) {
                string w = s.substr(right, one_word);
                right += one_word;
                if (m1.find(w) == m1.end()) {
                    count = 0;
                    left = right;
                    m2.clear();
                } else {
                    m2[w]++;
                    count++;
                    while (m2.at(w) > m1.at(w)) {
                        string tmp_w = s.substr(left, one_word);
                        m2[tmp_w]--;
                        count--;
                        left += one_word;
                    }
                }

                if (count == word_num) res.push_back(left);
            }
        }
        return res;
    }
};


int main() {
    string s = "barfoothefoobarman";
    vector<string> words = {"foo","bar"};
    Solution so;
    for (int i : so.findSubstring(s, words))
        cout << i << " ";
}

File: 030_Substring_with_Concatenation_of_All_Words/030_1.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.Queue;
import java.util.ArrayDeque;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Stack;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.Comparator;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if (s.isEmpty() || words.length == 0) return  res;
        int oneLen = words[0].length();
        int allLen = words.length * oneLen;
        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            int num = wordMap.getOrDefault(word, 0);
            wordMap.put(word, num + 1);
        }

        for (int i = 0; i < oneLen; i++) {
            int left = i, right = i, count = 0;
            Map<String, Integer> tmpMap = new HashMap<>();
            while (right + oneLen <= s.length()) {
                String w = s.substring(right, right + oneLen);
                right += oneLen;
                if (!wordMap.containsKey(w)) {
                    left = right;
                    count = 0;
                    tmpMap.clear();
                } else {
                    int num = tmpMap.getOrDefault(w, 0);
                    tmpMap.put(w, num + 1);
                    count++;
                    while (tmpMap.get(w) > wordMap.get(w)) {
                        String tmpWord = s.substring(left, left + oneLen);
                        tmpMap.put(tmpWord, tmpMap.getOrDefault(tmpWord, 0) - 1);
                        count--;
                        left += oneLen;
                    }
                }
                if (count == words.length) res.add(left);
            }
        }
        return res;
    }
}

public class leetcode {
    public static void main(String[] args) {
        String s = "barfoothefoobarman";
        String[] words = {"foo","bar"};
        Solution so = new Solution();
        System.out.println(so.findSubstring(s, words));
    }
}

